mirdata/dataset.py

```diff
@@ -0,0 +1,172 @@
+"""Module containing the Dataset base class
+"""
+import importlib
+import os
+import random
+
+import mirdata
+from mirdata import download_utils
+from mirdata import utils
+
+DATASETS = mirdata.__all__
+
+
+class Dataset(object):
+    """mirdata Dataset object
+
+    Usage example:
+    orchset = mirdata.Dataset('orchset')  # get the orchset dataset
+    orchset.download()  # download orchset
+    orchset.validate()  # validate orchset
+    track = orchset.choice_track()  # load a random track
+    print(track)  # see what data a track contains
+    orchset.track_ids()  # load all track ids
+
+    Attributes:
+        dataset (str): the identifier of the dataset
+        bibtex (str): dataset citation/s in bibtex format
+        remotes (dict): data to be downloaded
+        index (dict): dataset file index
+        download_info (str): download instructions or caveats
+        track_object (mirdata.track.Track): an uninstantiated Track object
+        dataset_dir (str): dataset save folder
+        readme (str): information about the dataset
+        data_home (str): path where mirdata will look for the dataset
+
+    """
+
+    def __init__(self, dataset, data_home=None):
+        """Inits a dataset by name and data location"""
+        if dataset not in DATASETS:
+            raise ValueError(
+                "{} is not a valid dataset in mirdata. Valid datsets are:\n{}".format(
+                    dataset, ",".join(DATASETS)
+                )
+            )
+        module = importlib.import_module("mirdata.{}".format(dataset))
+        self.dataset = dataset
+        self.bibtex = getattr(module, "BIBTEX", "")
+        self._remotes = getattr(module, "REMOTES", {})
+        self._index = module.DATA.index
+        self._download_info = getattr(module, "DOWNLOAD_INFO", None)
+        self._track_object = getattr(module, "Track", None)
+        self.dataset_dir = module.DATASET_DIR
+        self._download_fn = getattr(module, "_download", download_utils.downloader)
+        self.readme = module.__doc__
+
+        if data_home is None:
+            self.data_home = self.default_path
+        else:
+            self.data_home = data_home
+
+        # this is a hack to be able to have dataset-specific docstrings
+        self.track = lambda track_id: self._track(track_id)
+        self.track.__doc__ = self._track_object.__doc__  # set the docstring
+
+        # inherit any public load functions from the module
+        for method_name in dir(module):
+            if method_name.startswith("load_"):
+                method = getattr(module, method_name)
+                setattr(self, method_name, method)
```

Comment:
@magdalenafuentes I don't love this solution... but here's the problem - 
module-level functions weren't part of the dataset object, so if you wanted to use them, you'd have to both create an e.g. `b = mirdata.Dataset('beatles')` object and import the module, like `from mirdata import beatles` to do `beatles.load(b.choice_track.beat_path)`. Which is awful... so I put in this hack, which lists all variables and functions in the module (`dir(module)`) and adds the one starting with `load_` to the class. It works, but it feels very hacky.

The right way i think would be for all the loaders to inherit from a Dataset base class like this one. The top level (new) api could be similar `orchset = mirdata.dataset('orchset')`, but alllll of the loaders would change. The more I write this, the more I think it's the way we should go....and I'll probably work on this tomorrow :)

Reply:
I see what you mean. I think it makes sense that loaders inherit from this class and not "the other way around". This will be better in the long term as well, people could reuse this class to release their loaders directly with `mirdata-compliable` code

Reply:
I tried it, and it turns out to be much more complicated than it seemed :(. Buckle up for a long story 🙃 

**Option 1: (the one in the current PR) loaders do not extend the Dataset object**
The things i don't like about the current solution in this PR:
1) it's very hacky to import a module inside a class
2) the Track object needs data_home which should be set by the dataset object, so we have to define a function which is a simple wrapper on `Track` in order to call `dataset.track(track_id)` without repeating `data_home`.
3) because of 2, a hack (creating a lambda instead of a method & manually setting its docstring) is necessary for the dataset.track docstring to be specific to the dataset (and this is important to have)
4) manually attaching the module's load methods to the object is hacky

In summary...too many hacks 😸 . What is very nice about the solution in this PR is that adding a new loader is the same  as before (but much less!), and it's relatively clear what's going on. The top-level API is also as we want it.

To remove the hacks, I experimented with having loaders extend the Dataset class. And now the real story beings 📖 

**Option 2: loaders extend Dataset, and the load functions and the Track class are defined inside the Dataset object**
This addresses 1 and 4 but doesn't fix 2 and 3, and in fact makes 2 and 3 even uglier.

See the [dataset-object2 branch](https://github.com/mir-dataset-loaders/mirdata/blob/dataset-object2/mirdata/beatles.py#L215-L266) if you want to see the ugliness of the code. In short, the `Track` object needs access to the `load_*` functions, and `data_home`. Since `Track` is part of the Dataset class, it needs to access it's parent's `self`, which is hackily done in the base Class by [passing self as a variable](https://github.com/mir-dataset-loaders/mirdata/blob/dataset-object2/mirdata/dataset.py#L69-L71) to the `__init__` of `Track`. I looked and it's among the least ugly ways to do this - there are other ways but it's even worse. So now, to initialize a Track you have to call Track(track_id, data_home, dataset_object_self). Super weird, so the user should never see it, so to make it even uglier, we need to hide it by calling it `_Track`. Ugly right? Then, every time contributors are writing a loader, whenever they want to access a `load_*` function inside the `_Track` object, they have to call `self._dataset.load_*` instead of just calling `load_*`. Not the end of the world but not very friendly.

**Option 3: loaders extend Dataset but the load functions and the Track class are defined outside of Dataset**

This is kind of a mid-point between the current PR and Option 1. It would address 1 and 4 in the list of things I don't like, but doesn't solve 2 or 3. It adds two new problems - the users have to manually add the load_* functions to the Dataset object, and instead of defining globals (e.g. `REMOTES`) they have to set hidden variables in the init (e.g `self._remotes`). So it's a bit more copy-paste code than the current solution. 

```python
...
class Dataset(dataset.Dataset):
    def __init__(self, data_home):
       # everything here would be roughly the same as the current PR, but defined as attributes instead of globals
        self.bibtex = "..."
        self._remotes = {
            ...
        }
        ...
       # everything below this would be "extra" stuff the user writes compared to now
        self._track_object = Track
        self.load_beats = load_beats   # manually defined by the user
        self.load_sections = load_sections
        self.load_notes = load_notes
    ...

class Track(track.Track):
    ...

def load_beats(fpath):
    ...

def load_sections(fpath):
    ...

def load_notes(fpath):
    ...
```

Finally, for either of these solutions, we will still have to at some point hackily import a module in order to do something like `mirdata.get_dataset('orchset')` in the API. So we move the hacky import from inside the class to inside a function.

In summary, going back to my list of 4 hacks:
1. is fixed by options 2 or 3, but it moves the hacky import problem elsewhere, and it's not necessarily any better
2 and 3. are not fixed by any option, and are made worse by option 2
4. is fixed by option 2 at the cost of a lot of ugliness. it's also fixed by option 3, but at the cost of manual setting

So on points 1-3, I consider options 1 and 3 equivalent. On point 4, all options are bad in some way.

@magdalenafuentes now that you've read the novel, what do you think? 😂  Option 1, or option 3?

Reply:
and to bias you a bit, I am leaning towards option 1, because despite the hacks, the code is the shortest/cleanest in my opinion

Reply:
discussed with @magdalenafuentes and @drubinstein and we'll go with Option 1 for now, keeping option 3 in mind for the future if we find we need more flexibility. None of these are ideal, so we'll live with the hacks for now.